v3/lints/cabf_br/lint_subject_common_name_not_exactly_from_san.go

```diff
@@ -0,0 +1,69 @@
+package cabf_br
+
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type subjectCommonNameNotExactlyFromSAN struct{}
+
+/************************************************
+If present, this field MUST contain exactly one entry that is one of the values contained
+in the Certificate's `subjectAltName` extension
+
+If the [subject:commonName] is a Fully-Qualified Domain Name or Wildcard Domain Name, then
+the value MUST be encoded as a character-for-character copy of the dNSName entry value from
+the subjectAltName extension.
+************************************************/
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_subject_common_name_not_exactly_from_san",
+		Description:   "The common name field in subscriber certificates must include only names from the SAN extension",
+		Citation:      "BRs: 7.1.4.2.2",
+		Source:        lint.CABFBaselineRequirements,
+		EffectiveDate: util.CABFBRs_SC48_EffectiveDate,
+		Lint:          NewSubjectCommonNameNotExactlyFromSAN,
+	})
+}
+
+func NewSubjectCommonNameNotExactlyFromSAN() lint.LintInterface {
+	return &subjectCommonNameNotExactlyFromSAN{}
+}
+
+func (l *subjectCommonNameNotExactlyFromSAN) CheckApplies(c *x509.Certificate) bool {
+	return c.Subject.CommonName != "" && !util.IsCACert(c)
+}
+
+func (l *subjectCommonNameNotExactlyFromSAN) Execute(c *x509.Certificate) *lint.LintResult {
+	cn := c.Subject.CommonName
+
+	for _, dn := range c.DNSNames {
+		if cn == dn {
+			return &lint.LintResult{Status: lint.Pass}
+		}
+	}
+
+	for _, ip := range c.IPAddresses {
+		if cn == ip.String() {
```

Comment:
1. Doesn't Golang support `"<nil>"` as a value here? Should that be accounted for?
2. For IPv6, there's obviously several representations, and I missed that SC48 didn't put a normative dependency on to RFC 5952, specifically https://datatracker.ietf.org/doc/html/rfc5952#section-4 . Implicitly, I think this logically holds, but that's not explicitly stated anywhere. At the risk of significantly overthinking things, do you think we should add an extra pass that, if there are no exact matches (which pass), we do a fuzzy match for IPv6 representations (I guess attempting to parse the CN to an IP, then seeing if that exactly matches any of the raw/non-stringified values), and warn if it's a fuzzy match (aka a different string encoding)

Reply:
@CBonnell rightfully pointed out I can't read - and https://github.com/cabforum/servercert/pull/285/files#diff-e0ac1bd190515a4f2ec09139d395ef6a8c7e9e5b612957c1f5a2dea80c6a6cfeR2092 is the requirement

That said, can you confirm Golang's ip.String() follows that RFC?

Reply:
So, according to the comment on the IP.String() function:
```
// String returns the string form of the IP address ip.
// It returns one of 4 forms:
//   - "<nil>", if ip has length 0
//   - dotted decimal ("192.0.2.1"), if ip is an IPv4 or IP4-mapped IPv6 address
//   - IPv6 ("2001:db8::1"), if ip is a valid IPv6 address
//   - the hexadecimal form of ip, without punctuation, if no other cases apply
```
So I think it's probably just the latter case we want to think about catching?

Reply:
I've been thinking and I'm inclined to think that if we want to check that the SAN:IPAddresses are valid we should be doing that using a separate lint to split the CN-in-SAN check from the IPAddresses validity check.

Reply:
Right, the issue here isn't whether the SAN:IPAddresses are valid - those are stored in an unambiguous network bytes. The issue is the stringified form here. What I wanted to make sure is that the stringification function we're using (here, `IP.String()`) matches the rules set out in SC48.

The problem is that there are edge cases here:
- For not v4/v6 addresses, it prepends with a `?` (the documentation doesn't mention this, but the implementation does do this)
  - This isn't so much a concern, because SAN will always be v4 or v6, and that's handled by the X509 parser
- For IPv6, it appears to correctly implement `4.1` - `4.3` of https://datatracker.ietf.org/doc/html/rfc5952#section-4
- However, it does not correctly handle Section 5 - for example, an input of "::ffff:192.0.2.1" returns an output of "192.0.2.1" - https://play.golang.org/p/T7Cd-mWqZ3R
  - To be fair, I'm not sure the expected result here for IPv4-mapped addresses (from the ballot), but maybe this is what you meant by a separate lint to tackle these (e.g. should a v4 mapped address be stored *as* a v4 SAN?)

Reply:
SC48 clarified that IPv4-mapped addresses are considered to be Reserved IP addresses and should not be present in Certificates (the definition of Reserved IP address now states that all address blocks listed in this IANA registry here are considered Reserved: https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml). IPv4-mapped addresses are entry #3 in the registry. Given this, I don't think we need to be concerned with @sleevi's last corner case.

Reply:
Thanks for the reply @CBonnell.  For the record, the reserved IP ranges are checked by lint "e_subject_contains_reserved_ip" so I believe we can consider this resolved.

Reply:
I will defer further reviews until after I have both coffee in hand and have fully reviewed all of SC48, and not just looking at implementation edge cases with stringifying functions ;)

Reply:
@CBonnell Thanks for https://github.com/zmap/zlint/pull/627/files#r695901667 - but that's not presently tested for in any lint, AFAICT. https://github.com/zmap/zlint/blob/d5ab97e9ca1cef2c7b594672c0265cef3b703637/v3/util/ip.go#L90-L114 contains the list of reserved addresses, which is what the `lint_ext_san_contains_reserved_ip` uses to implement its test.

Looking at the test, neither Golang's `ip.IsGlobalUnicast` nor `reservedNetworks` covers the IPv4-mapped IPv6 addresses. Presumably, this is because Golang internally tests `ip.To4()` when performing a number of these checks, while the `util/ip.go` does not.

Originally I was just going to suggest that we add a comment mentioning that IPv4-mapped IPv6 addresses are already accounted for via `subject_contains_reserved_ip`, but at present, that's not the case.

@robplee I've had my coffee (and then some), but I'm still not sure I've fully traced through all this path. You could certainly empirically test, but I _suspect_ we may need to add `::ffff:0:0/96` to the reserved networks. This isn't quite precise from an "IP protocol" layer (since it's really treated as a v4 address smuggled in a v6 length), but for purposes of zlint, it's totally correct, because we shouldn't be smuggling such mapped addresses as IPv6 SANs. If you do that, then perhaps adding a comment here noting the edge case of `IP.ToString()`

```suggestion
		// IPv4-mapped IPv6 addresses (e.g. "::ffff:0:0:192.168.0.1") are converted to their IPv4 equivalent
		// ("192.168.0.1") by IP.String(). This is not handled here, as it's treated as a reserved addresses in
		// util.IsIANAReserved, and will be rejected by cabf_br/lint_ext_san_contains_reserved_ip.
		if cn == IP.String() {
```

Reply:
I think this is the only open question remaining and unless you've had time to look further and determine strict confirmation either way then I can add the changes you recommend of putting "::ffff:0:0/96" into the list of reserved networks and add the comment text you proposed in also and we should be covered against those concerns.